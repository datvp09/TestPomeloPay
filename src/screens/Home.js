import React, {useEffect, useState} from 'react';
import {
  StyleSheet,
  View,
  Text,
  FlatList,
  Image,
  TouchableOpacity,
} from 'react-native';
import axios from 'axios';
import {connect} from 'react-redux';
import Modal from 'react-native-modal';
import {NavigationHeader, Toast} from '../components';
import {refundIcon, intenseBlue, isIpX} from '../Constants';
import moment from 'moment';

const mockupURL =
  'https://gist.githubusercontent.com/sverraest/7be1341f3a92391edf629c09c8749d15/raw/ee87f5c1722b9b11198a35cba5fd9d068135adf4/gistfile1.txt';
const REFUNDED_MSG = 'REFUNDED';
const CONFIRMED_MSG = 'CONFIRMED';
const QRCODEGENERATED_MSG = 'QR_CODE_GENERATED';

const Home = ({transactions}) => {
  const [isLoading, setIsLoading] = useState(false);
  const [isRefundAlertShow, setRefundVisible] = useState(false);

  useEffect(() => {
    onRefresh();
  }, []);

  const showRefundAlert = () => setRefundVisible(true);

  const hideRefundAlert = () => setRefundVisible(false);

  const onRefundAllConfirmed = () => {
    const refundedTransactions = transactions.map(x => ({
      ...x,
      state: 'REFUNDED',
    }));
    setTransactions(refundedTransactions);
    hideRefundAlert();
  };

  const renderRightView = () => (
    <TouchableOpacity onPress={showRefundAlert} style={styles.refundAllButton}>
      <Image
        source={refundIcon}
        resizeMode={'contain'}
        style={styles.refundAllButtonImg}
      />
    </TouchableOpacity>
  );

  const renderItems = ({item, index}) => {
    const {
      amount,
      currency,
      created, // Date object
      expires, // Date object
      id,
      initiatorDetails: {contactEmail, contactName},
      state,
    } = item;
    let stateColor = 'green'; // CONFIRMED_MSG
    if (state == REFUNDED_MSG) {
      stateColor = 'red';
    } else if (state == QRCODEGENERATED_MSG) {
      stateColor = intenseBlue;
    }

    return (
      <View style={[styles.itemTitleRow, index == 0 && {marginTop: 10}]}>
        <Text style={styles.itemTitle}>{'#' + id}</Text>
        <View style={styles.labelWrap}>
          <Text style={styles.itemText}>
            {'Amount: ' + amount + ' ' + currency}
            <Text style={{color: stateColor, fontWeight: '600'}}>
              {'\t\t' + state}
            </Text>
          </Text>
          <Text style={styles.itemText}>{'Contact name: ' + contactName}</Text>
          <Text style={styles.itemText}>
            {'Contact email: ' + contactEmail}
          </Text>
          <Text style={styles.itemText}>
            {'Create date: ' + moment(created).format('DD-MM-YYYY')}
          </Text>
          <Text style={styles.itemText}>
            {'Expire date: ' + moment(expires).format('DD-MM-YYYY')}
          </Text>
        </View>
      </View>
    );
  };

  const onRefresh = () => {
    setIsLoading(true);
    axios(mockupURL)
      .then(res => {
        if (res?.status != 200) {
          Toast('Error fetching data');
          return;
        }
        const newData = [...transactions];
        res?.data?.items.forEach(x => {
          if (transactions.some(y => y.id == x.id)) {
            newData.push(x);
          }
        });
        setTransactions(newData);
      })
      .catch(e => {
        Toast('Error fetching data');
        setTransactions([]);
      })
      .finally(() => setIsLoading(false));
  };

  return (
    <>
      <View style={styles.screen}>
        <NavigationHeader
          style={{height: 60}}
          title={'All Transactions'}
          rightView={renderRightView}
        />
        <FlatList
          data={transactions}
          keyExtractor={item => `${item.id}`}
          renderItem={renderItems}
          style={styles.list}
          refreshing={isLoading}
          onRefresh={onRefresh}
        />
        <Modal
          isVisible={isRefundAlertShow}
          animationIn="fadeIn"
          animationOut="fadeOut"
          backdropTransitionOutTiming={0}
          onBackButtonPress={hideRefundAlert}
          onBackdropPress={hideRefundAlert}>
          <View style={styles.alertContainer}>
            <Text style={styles.refundConfirmText}>
              {'Confirm refund all transactions'}
            </Text>
            <TouchableOpacity
              style={styles.refundConfirmButton}
              onPress={onRefundAllConfirmed}>
              <Text>{'OK'}</Text>
            </TouchableOpacity>
          </View>
        </Modal>
      </View>
    </>
  );
};

const styles = StyleSheet.create({
  screen: {
    flex: 1,
    backgroundColor: '#d0d0d0',
  },
  list: {
    flex: 1,
    marginBottom: isIpX ? 40 : 10,
  },
  itemTitleRow: {
    marginVertical: 5,
    marginHorizontal: 7,
    padding: 12,
    backgroundColor: 'white',
    borderRadius: 4,
    marginBottom: 5,
  },
  alertContainer: {
    height: 100,
    width: '80%',
    alignSelf: 'center',
    justifyContent: 'space-between',
    backgroundColor: 'white',
    borderRadius: 4,
  },
  refundAllButton: {
    marginRight: 15,
    paddingTop: 3,
    width: 40,
    height: 30,
    alignItems: 'center',
  },
  refundAllButtonImg: {
    width: 20,
    height: 20,
    tintColor: 'white',
  },
  refundConfirmText: {
    flex: 1,
    marginTop: 15,
    textAlign: 'center',
  },
  refundConfirmButton: {
    width: '100%',
    height: 35,
    justifyContent: 'center',
    alignItems: 'center',
    borderTopWidth: 1,
    borderTopColor: '#aaa',
    borderRadius: 4,
    backgroundColor: 'white',
  },
});

const mapStateToProps = state => ({
  transactions: state.transaction.transactions,
});

const mapDispatchToProps = dispatch => {
  const {setTransactions} = require('../../redux/actions');

  return {
    setTransactions: data => dispatch(setTransactions(data)),
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(Home);
