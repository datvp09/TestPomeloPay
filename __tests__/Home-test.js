import 'react-native';
import React from 'react';
import Home from '../src/screens/Home';
import renderer from 'react-test-renderer';
import {Provider} from 'react-redux';
import configureMockStore from 'redux-mock-store';

const mockStore = configureMockStore();
const store = mockStore({
  transaction: {
    transactions: [],
  },
});

describe('Home', () => {
  describe('Rendering', () => {
    it('should match to snapshot', () => {
      const component = renderer
        .create(
          <Provider store={store}>
            <Home />
          </Provider>,
        )
        .toJSON();
      expect(component).toMatchSnapshot();
    });
  });
});
