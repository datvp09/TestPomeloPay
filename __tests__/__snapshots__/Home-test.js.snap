// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Rendering should match to snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#d0d0d0",
      "flex": 1,
    }
  }
>
  <View>
    <View
      style={
        Object {
          "backgroundColor": "#1365AF",
          "height": 20,
        }
      }
    />
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#1365AF",
            "flexDirection": "row",
            "height": 55,
            "justifyContent": "space-between",
            "paddingLeft": 25,
          },
          Object {
            "height": 60,
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "white",
              "fontSize": 17,
            },
            undefined,
          ]
        }
      >
        All Transactions
      </Text>
      <View
        accessible={true}
        focusable={true}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "height": 30,
            "marginRight": 15,
            "opacity": 1,
            "paddingTop": 3,
            "width": 40,
          }
        }
      >
        <Image
          resizeMode="contain"
          source={
            Object {
              "testUri": "../../../src/assets/images/iconBack.png",
            }
          }
          style={
            Object {
              "height": 20,
              "tintColor": "white",
              "width": 20,
            }
          }
        />
      </View>
    </View>
  </View>
  <RCTScrollView
    data={Array []}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onRefresh={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    refreshControl={
      <RefreshControlMock
        onRefresh={[Function]}
        refreshing={false}
      />
    }
    refreshing={false}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    style={
      Object {
        "flex": 1,
        "marginBottom": 10,
      }
    }
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <RCTRefreshControl />
    <View />
  </RCTScrollView>
  <Modal
    animationType="none"
    hardwareAccelerated={false}
    hideModalContentWhileAnimating={false}
    onModalHide={[Function]}
    onModalWillHide={[Function]}
    onModalWillShow={[Function]}
    onRequestClose={[Function]}
    scrollHorizontal={false}
    scrollOffset={0}
    scrollOffsetMax={0}
    scrollTo={null}
    supportedOrientations={
      Array [
        "portrait",
        "landscape",
      ]
    }
    swipeThreshold={100}
    transparent={true}
    visible={false}
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "black",
          "bottom": 0,
          "height": 1334,
          "left": 0,
          "opacity": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
          "width": 750,
        }
      }
    />
    <View
      hideModalContentWhileAnimating={false}
      onModalHide={[Function]}
      onModalWillHide={[Function]}
      onModalWillShow={[Function]}
      pointerEvents="box-none"
      scrollHorizontal={false}
      scrollOffset={0}
      scrollOffsetMax={0}
      scrollTo={null}
      style={
        Object {
          "flex": 1,
          "justifyContent": "center",
          "margin": 37.5,
          "transform": Array [
            Object {
              "translateY": 0,
            },
          ],
        }
      }
      supportedOrientations={
        Array [
          "portrait",
          "landscape",
        ]
      }
      swipeThreshold={100}
    >
      <View
        style={
          Object {
            "alignSelf": "center",
            "backgroundColor": "white",
            "borderRadius": 4,
            "height": 100,
            "justifyContent": "space-between",
            "width": "80%",
          }
        }
      >
        <Text
          style={
            Object {
              "flex": 1,
              "marginTop": 15,
              "textAlign": "center",
            }
          }
        >
          Confirm refund all transactions
        </Text>
        <View
          accessible={true}
          focusable={true}
          isTVSelectable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "white",
              "borderRadius": 4,
              "borderTopColor": "#aaa",
              "borderTopWidth": 1,
              "height": 35,
              "justifyContent": "center",
              "opacity": 1,
              "width": "100%",
            }
          }
        >
          <Text>
            OK
          </Text>
        </View>
      </View>
    </View>
  </Modal>
</View>
`;
